python/lsst/pipe/base/graph/_implDetails.py

```diff
@@ -181,3 +233,119 @@ def __contains__(self, key: _T) -> bool:
+    r"""Prunes supplied dataset refs out of datasetRefDict container, recursing
+    to additional nodes dependant on pruned refs. This function modifies
+    datasetRefDict in-place.
+
+    Parameters
+    ----------
+    datasetRefDict : `_DatasetTracker[DatasetRef, QuantumNode]`
+        The dataset tracker that maps `DatasetRef`\ s to the Quantum Nodes
+        that produce/consume that `DatasetRef`
+    refsToRemove : `Iterable` of `DatasetRef`
+        The `DatasetRef`\ s which should be pruned from the input dataset
+        tracker
+    alreadyPruned : `set` of `QuantumNode`
+        A set of nodes which have been pruned from the dataset tracker
+    """
+    if alreadyPruned is None:
+        alreadyPruned = set()
+    for ref in refsToRemove:
+        # make a copy here, because this structure will be modified in
+        # recursion, hitting a node more than once won't be much of an
+        # issue, as we skip anything that has been processed
+        nodes = set(datasetRefDict.getConsumers(ref))
+        for node in nodes:
+            # This node will never be associated with this ref
+            datasetRefDict.removeConsumer(ref, node)
+            if node in alreadyPruned:
+                continue
+            # find the connection corresponding to the input ref
+            connectionRefs = node.quantum.inputs.get(ref.datasetType)
+            if connectionRefs is None:
+                # look to see if any inputs are component refs that match the
+                # input ref to prune
+                others = ref.datasetType.makeAllComponentDatasetTypes()
+                # for each other component type check if there are assocated
+                # refs
+                for other in others:
+                    connectionRefs = node.quantum.inputs.get(other)
+                    if connectionRefs is not None:
+                        # now search the component refs and see which one
+                        # matches the ref to trim
+                        for cr in connectionRefs:
+                            if cr.makeCompositeRef() == ref:
+                                toRemove = cr
+                        break
+                else:
+                    # Ref must be an initInput ref and we want to ignore those
+                    raise RuntimeError(f"Cannot prune on non-Input dataset type {ref.datasetType.name}")
+            else:
+                toRemove = ref
+
+            tmpRefs = set(connectionRefs).difference((toRemove,))
+            tmpConnections = NamedKeyDict(node.quantum.inputs.items())
+            tmpConnections[toRemove.datasetType] = list(tmpRefs)
+            helper = AdjustQuantumHelper(inputs=tmpConnections, outputs=node.quantum.outputs)
+            msg = ("assert to make the type checker happy, it should not actually be possible to not "
+                   "have dataId set to None at this point")
+            assert node.quantum.dataId is not None, msg
```

Comment:
Assigning the assert message to a variable outside the assert statement seems a bit against the spirit of asserts and being able to "compile them out".